.circleci/config.yml

```diff
@@ -266,43 +267,85 @@ jobs:
           name: Upload Coverage
           command: bash <(curl -s https://codecov.io/bash) -cF python -s "/tmp/workspace/coverage/"
 
-  build_image:
+  build_docker_image:
     docker:
       - image: docker
+    parameters:
+      python_version:
+        type: string
+      tag_latest:
+        type: boolean
+        default: false
+    environment:
+      PYTHON_VERSION: << parameters.python_version >>
+      PYTHON_TAG: python<< parameters.python_version >>
     steps:
-      - setup_remote_docker
       - checkout
       - run:
-          name: Docker Build
-          command: docker build -t prefecthq/prefect .
+          # todo: is there a better way to ensure that this is a commit on master?
+          name: Master branch check
+          command: |
+            apk add git
+            if [[ $(git branch --contains $CIRCLE_SHA1 --points-at master | wc -l) -ne 1 ]]; then
+              echo "commit $CIRCLE_SHA1 is not a member of the master branch"
+              exit 1
+            fi
+      - setup_remote_docker:
+          docker_layer_caching: true
+      - run:
+          name: Build image
+          command: >-
+            docker build 
+            --build-arg GIT_SHA=$CIRCLE_SHA1 
+            --build-arg BUILD_DATE=$(date -u +'%Y-%m-%dT%H:%M:%SZ')
+            --build-arg PREFECT_VERSION=$CIRCLE_TAG 
+            -t prefecthq/prefect:${CIRCLE_TAG}-${PYTHON_TAG}
```

Comment:
whenever `CIRCLE_TAG` is not populated (e.g., on a normal merge-to-master), will this be a valid image tag?

Reply:
no it will not, I believe it would be an empty or missing value. That is the reason for [the filter](https://github.com/PrefectHQ/prefect/pull/1648/files#diff-1d37e48f9ceff6d8030570cd36286a61R364), such that we are only calling this job upon a tagged commit.

However, I could add a `set -u` before this command such that bash will throw an error upon attempting to access a non-existent or empty variable?